### ISSUE TITLE
SupersetDB never recovers out of failed state / Remove SupersetDB mechanism

### ISSUE THREAD

**Issue by sbernauer** [stackabletech/superset-operator#351]
Reported in https://github.com/stackabletech/superset-operator/issues/347
Once failed the state machine can never recover from the failed state: https://github.com/stackabletech/superset-operator/blob/355965e0517ec81f22b285be9f16c407e116bdf2/rust/operator-binary/src/superset_db_controller.rs#L221

My suggestion is to actually get rid of the SupersetDB. Maybe can simply have an init-container which prepares the DB and call it a day. Kind of similar to what hive-op is doing.

Problems with current SupersetDB approach:
* It caused problems in the past, such as support questions from users e.g. wiping/re-creating the postgres and the Superset tables not getting created because the SupersetDB says "everything is fine". Users had a really hard time understanding what's going on
* Users have changed the admin password, change did not propagate
* Think of Superset version bumpes: The superset-op needs to track the current version as well as the previous version and set the SupersetDB back to uninitialized, so that the k8s Job bumps the tables in postgres. Things get even worse when the k8s job still exists and the SupersetDB controller tries to patch unmutable fields on the k8s job. Lot's of pain.
Sebastian from the future here: We actually run into this problem during bump Superset 1.5 -> 2.0

**Comment by sbernauer** [comment]
Running into problems again: https://stackable-workspace.slack.com/archives/C04CUN2RS31/p1681726447481479
(docker image not updated because of AirflowDB not chaning)

**Comment by fhennig** [comment]
I think it's worth maybe thinking about redesigning this. But all operators need to be idempotent. I'm not sure if i.e. the example loading is idempotent too, or if it will overwrite and take a long time every time it starts. Just saying we should check.

The CRD keeps track of this status currently.

**Comment by sbernauer** [comment]
@sbernauer did run into this with a customer setup while upgrading Superste
@Maleware did also run into this problem with another customer

**Comment by sbernauer** [comment]
Another customer has run into this (with Airflow)

**Comment by dervoeti** [comment]
We just discussed this problem, there are basically two options:

1. Remove SupersetDB and always init/upgrade the database in an init container of the Superset Pod
2. Keep SupersetDB and handle database upgrades using a one-time job, if the version changes

Pros of the first option:
  - Simpler for us, less scenarios we have to think about
  - Probably less confusing for users

Cons of the first option:
  - Not sure if the whole init process is idempotent
  - Not sure if there are any problems when the init runs in parallel (multiple Superset Pods starting at the same time)
  - Longer startup time for the Superset Pod

Pros of the second option:
  - It's somehow cleaner: A db upgrade should only be run once, when upgrading to an new Superset version, not every time a Pod starts
  - Maybe we could start small by just handling DB upgrades and doing a one-time init on creation, then extend the functionality later (automatic updates when admin password is changed etc.)
  - Faster startup time for the Superset Pod

Cons of the second option:
  - More complex for us, more scenarios we have to think about
  - Maybe more confusing for users

For now, we decided to try the first option and see what problems we run into. We'll also ask the Superset Mailing List for their opinion.

**Comment by dervoeti** [comment]
Alright, I tested a lot of scenarios, short resume: Option 1 basically works fine.
We should introduce liveness/readiness probes, not sure why they were not present for Superset (they are for example present for the Airflow webserver). Also, I think we should change the `podManagementPolicy` to `OrderedReady`, which is the default. Right now its set to `Parallel`. That would make sure multiple replicas start up correctly after another.
Then there is one problem left:
Loading the examples multiple times can fail in certain situations (for example, when upgrading Superset from 1.5.3 to 2.1.0 and then trying to load the examples). If we somehow make sure the examples are only loaded once when the cluster is created, option 1 should be totally fine.
I can think of three solutions for this:
 - Create a one-time Job for loading the examples
 - Find a way to check if the examples are already loaded (maybe by checking the database)
 - Run "superset load_examples || true", so basically just ignore errors when loading the examples
Any opinions on this?

**Comment by sbernauer** [comment]
Thanks for the research, awesome!
As I would not expect any production setup to use examples, I would be fine with the least complex and expensive solution. Especially as we should be able to non-breaking switch to a different way if we discover problems.
Hence Option "superset load_examples || true" with the possibility to switch to something different in the future would be fine for me ^^

**Comment by fhennig** [comment]
Yes, thank you very much for your reasearch!

After giving this some thought I came up with an idea: How about we get rid of the example loading feature completely, but instead we make a demo that runs a custom job to load the examples? That would fit much better into our other examples/demos and this example loading feature is out of the CRD as well. And the demos are completly under our control and not part of the strict non-breaking processes etc. that we have with the CRD. We could just ditch the demo in the future if we don't want to support it anymore.

So it's a combination of "Remove support for loading the examples" and "Create a one-time Job for loading the examples"

**Comment by sbernauer** [comment]
Would be ok for me as well. I like the `loadExamples` as it can be enabled after the fact once you already have a running Superset installation. Like we have setup Superset for customer X, he asks: How can I do a geojson chart, and we load the examples for him so he can have a look.

**Comment by dervoeti** [comment]
I like the idea of getting rid of the `loadExamplesOnInit` option completely. It's not really a setting like "use these credentials/resources/listener addresses". To me it's not really "cluster configuration". So I'd agree it should be in a demo, not in the CRD. Or at least document how a user can load the examples, if we don't want to create a demo.

But that's only the technical side of it. On the other hand, it's more comfortable for a customer to just set that setting to `true` to load the examples. Not sure how often customers really make use of this though.

**Comment by sbernauer** [comment]
Having a docs section where the user `kubectl appy`s a Job that loads the examples sounds good to me. If enough users are doing this we can re-add dedicated support for this.
But I'm also happy to implement it "propertly" via a CRD field

**Comment by dervoeti** [comment]
Yeah, documentation with a ready-to-go `kubectl apply` command is probably comfortable enough. That's maybe even easier for the customer than modifying the CRD.
And we get rid of the responsibility to manage the loading of examples. I fear that `superset load_examples || true` might cause problems in the future, for example: loading the examples fails and leaves the cluster in a broken state.
It also reduces the initial startup time of the cluster, the customer doesn't have to wait for the examples to load, because it runs in the background as a Job. Loading the examples usually takes several minutes.

**Comment by fhennig** [comment]
> I like the idea of getting rid of the loadExamplesOnInit option completely. It's not really a setting like "use these credentials/resources/listener addresses". To me it's not really "cluster configuration". So I'd agree it should be in a demo, not in the CRD. Or at least document how a user can load the examples, if we don't want to create a demo.

> Having a docs section where the user kubectl appys a Job that loads the examples sounds good to me. If enough users are doing this we can re-add dedicated support for this.

that's great, sounds like we all agree!

> And we get rid of the responsibility to manage the loading of examples. I fear that superset load_examples || true might cause problems in the future, for example: loading the examples fails and leaves the cluster in a broken state.
It also reduces the initial startup time of the cluster, the customer doesn't have to wait for the examples to load, because it runs in the background as a Job. Loading the examples usually takes several minutes.

I quite like this part of it as well, it's something I considered too.

So then i'd propose:

- remove the field 
- add a docs page that documents how to load the examples with a copy-pastable command that runs a job to load them (where this will go exactly is still TBD)

**Comment by dervoeti** [comment]
Closed by https://github.com/stackabletech/superset-operator/pull/396 (code) and https://github.com/stackabletech/superset-operator/pull/398 (docs)

**Comment by lfrancke** [comment]
Does this change the CRD?
Or more specifically (I didn't look at the details): This requires a migration, right?

**Comment by fhennig** [comment]
Ah, good point. A config option was removed. So I think yes, this is a breaking change and requires a migration?

**Comment by maltesander** [comment]
Yeah this is breaking. Two optional fields and the DB CRD were removed. Migration looks harder than it is since the complexity was reduced alot and a full standalone CRD removed. This should be cleaned up though.

**Comment by dervoeti** [comment]
I'm not sure if this is considered breaking. The `.spec.clusterConfig.loadExamplesOnInit` option was removed from the CRD and the `SupersetDB` CR is not needed any more. So customers just can't use the `loadExamplesOnInit` option anymore for new clusters and can remove the `SupersetDB` resources (as well as the CRD) from existing clusters. So it only breaks when customers use a new version of the CRD with an old definition of a `SupersetCluster` that has `loadExamplesOnInit` set. When they just upgrade the operator, everything should work fine (CRD is not upgraded). And if they upgrade the CRD, the `loadExamplesOnInit` should automatically be removed from existing CRs (without any side effects).

I sketched out something for the release notes:
https://github.com/stackabletech/documentation/blob/release/23.11/modules/ROOT/pages/release_notes.adoc
This is more a reminder than a final version. We will implement roughly the same change for Airflow, this will also need to be included in the release notes, maybe we can merge some of those paragraphs.
Not sure if we should include a one-liner to remove the SupersetDB/AirflowDB CRs (like `kubectl get ... | grep ... | xargs kubectl delete ...`) in the release notes as well.
As far as I see it, including the change in the release notes is probably the only thing left to do with regards to migration.
